prboom2/src/dsda/time.c

```diff
@@ -15,58 +15,37 @@
 //	DSDA Time
 //
 
+#ifdef _WIN32
+#include <windows.h>
+#else
 #include <time.h>
+#endif
 
 #include "i_system.h"
 
 #include "time.h"
 
-// [XA] MSVC hack: need to implement a handful of functions here.
-// this was yoinked from https://stackoverflow.com/a/31335254 so
-// no guarantees on if this stuff is accurate/kosher/whatever,
-// so maybe don't cut an official release with MSVC just yet.
-// it's enough to hack it into compiling for me, at least. ;)
-
-#ifdef _MSC_VER
-
-#include <windows.h>
+// lovey: Win32 interface
+#ifdef _WIN32
 
-#define CLOCK_MONOTONIC -1
-#define exp7           10000000i64 //1E+7
-#define exp9         1000000000i64 //1E+9
-#define w2ux 116444736000000000i64 //1.jan1601 to 1.jan1970
+static unsigned long long dsda_time[DSDA_TIMER_COUNT];
 
-void unix_time(struct timespec *spec) {
-  __int64 wintime; GetSystemTimeAsFileTime((FILETIME*)&wintime);
-  wintime -= w2ux;  spec->tv_sec = wintime / exp7;
-  spec->tv_nsec = wintime % exp7 * 100;
+void dsda_StartTimer(int timer) {
+  QueryPerformanceCounter((LARGE_INTEGER*)&dsda_time[timer]);
 }
 
-int clock_gettime(int _skip, struct timespec *spec)
-{
-  static struct timespec startspec;
-  static double ticks2nano;
-  static __int64 startticks, tps = 0;
-  __int64 tmp, curticks;
-
-  QueryPerformanceFrequency((LARGE_INTEGER*)&tmp);
+unsigned long long dsda_ElapsedTime(int timer) {
+  static unsigned long long freq = 0;
+  unsigned long long now;
 
-  if (tps != tmp) {
-    tps = tmp;
-    QueryPerformanceCounter((LARGE_INTEGER*)&startticks);
-    unix_time(&startspec); ticks2nano = (double)exp9 / tps;
-  }
+  if (!freq) QueryPerformanceFrequency((LARGE_INTEGER*)&freq);
 
-  QueryPerformanceCounter((LARGE_INTEGER*)&curticks); curticks -= startticks;
-  spec->tv_sec = startspec.tv_sec + (curticks / tps);
-  spec->tv_nsec = startspec.tv_nsec + (double)(curticks % tps) * ticks2nano;
-  if (!(spec->tv_nsec < exp9)) { spec->tv_sec++; spec->tv_nsec -= exp9; }
-  return 0;
+  QueryPerformanceCounter((LARGE_INTEGER*)&now);
+  return (now-dsda_time[timer])*1000000ll/freq;
```

Comment:
Where does this constant come from? Can you give it a name?

Reply:
The unix code doesn't specify what it means in this context either, so I didn't think it mattered.

It's converting the "counts" (what you get from `QueryPerformanceCounter`) to microseconds.

The frequency of these "counts" is queried from `QueryPerformanceFrequency`, which is what freq is filled with.